src/H5O.c

```diff
@@ -1253,6 +1426,102 @@ H5Oclose(hid_t object_id)
     FUNC_LEAVE_API(ret_value)
 } /* end H5Oclose() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5Oclose
+ *
+ * Purpose:     Close an open file object.
+ *
+ *              This is the companion to H5Oopen. It is used to close any
+ *              open object in an HDF5 file (but not IDs are that not file
+ *              objects, such as property lists and dataspaces). It has
+ *              the same effect as calling H5Gclose, H5Dclose, or H5Tclose.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ * Programmer:	James Laird
+ *		July 14 2006
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5Oclose_async(const char *app_file, const char *app_func, unsigned app_line, 
+    hid_t object_id, hid_t es_id)
+{
+    H5VL_object_t *vol_obj = NULL;  /* Object for loc_id */
+    H5VL_t *connector = NULL;       /* VOL connector */
+    void *token = NULL;             /* Request token for async operation        */
+    void **token_ptr = H5_REQUEST_NULL; /* Pointer to request token for async operation        */
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_API(FAIL)
+    H5TRACE5("e", "*s*sIuii", app_file, app_func, app_line, object_id, es_id);
+
+    /* Get the type of the object and close it in the correct way */
+    switch (H5I_get_type(object_id)) {
```

Comment:
Could this switch statement with the "is it OK to close this ID" check be moved into common code, for both H5Oclose and H5Oclose_async to use?

Reply:
The other code here (and in H5Oclose) would be clearer, after that.